### PR TITLE
Replace legacy tvdb_api with tvdb api v2

### PR DESCRIPTION
**Pull request by p0psicles**
### The goals of this PR are:
1. Restructure the indexer code, creating a more modular setup, where it's less work to add new indexers compared to how it is currently setup.

2. Replace the 'legacy' thetvdb indexer code with the api v2. https://api.thetvdb.com/swagger.
All tough allot the added api code is now in this PR, it will be added as a separate PR, as the api, should be fully independent from the indexer's code. This probably the biggest change from this PR. As up until now, api and indexer where pretty much the same.

A followup will be adding the tvmaze indexer. As is waiting in #145 

### To accomplish this the following has to be done:
- [x] Make sure the swagger client has all the features as the legacy client.
- [x] Separate the indexer code into a lib part and medusa indexer code. Currently they're tangled.
- [x] Add the v2 code as a new indexer next to the legacy indexer. The legacy indexer code will be replaced. But for testing purposses, they should work fine next to each other. The ID used will be '4' for now, as 3 has been reserved by tvmaze and '2' is used by tvrage.
- [x] When tested, remove the legacy indexer code, and make sure everything works fine, including the scene names / season scene names / scene episode mappings.

### Log:
This is a list to document what needs to be taken care of when implementing a new indexer. I'll update this as I move along.

- [x] Show search by name
- [x] Show search by id, indexing the show attributes, seasons, episodes, banner, etc. `Is working, except for the banner.`
- [x] Episode summaries
- [x] Add metadata creation
- [x] Add banner and poster support if available
- [x] Add actors information (Used for metadata creation)
- [x] Remove all references to tvdb legacy indexer. The api, might also be called directly.
- [x] Add multilingual support, (seach shows, get show data, get episode data and get images in the locale configured).
- [x] Add timezone and network information
- [ ] Add indexer mappings (when indexer providers indexer ids to other indexers)
- [x] Add the indexer icon
- [ ] Lot's of cleanup of the tvdbv2 indexer files (pep8, snake_case, unicode_literals)
- [ ] Add new unit tests @ratoaq2 could really need some help here.
- [x] Implement a new method for show updating, as we cannot use the old tvdb xml's anymore. This will need allot of testing and refinement!

### Tvdb Apiv2 Lib related
- [x] Make sure the api returns unicode. Test by searching: 'guide to pretty'. Currently i've adjusted the mapping for the Series modal. Need to also check for others, like episode, actors, etc.
- [x] Maybe replace urllib3 with requests. But the way it has been setup, that might not be that easy. Not needed, as for urllib3, we dont need to install any additional libs.
- [ ] Add searching with apostophes '.

### showUpdates related
More info: https://github.com/pymedusa/Medusa/wiki/Show-indexer-updating
For this the following needs to be implemented:
- [x] Support for force full update
- [x] Update only seasons that reached their next_update.
- [x] Add new shows, should set the next_updates
- [x] Episodes that start a new season, should update the cache indexer_update table with the next_update.
@fernandog, did you have some other use cases/flows?

### Screenies
v2 api is already giving more results back than the old one.
image

v1:
image

Update 01/10: From the list of unchecked bullets, it looks like this PR is far from finished. There is still allot of work/reviewing to be done. But this one is already at like 80%. With the major features in it.

### Legacy, places where TVDB constant is used. Need to check  these. I'm creating a note for it, because if were moving to multi indexer in the future, these need to go.
* /medusa/databases/main_db.py
* /medusa/providers/torrent/json/rarbg.py
* /medusa/scene_exceptions.py
* /medusa/server/web/home/handler.py
* /medusa/tv.py
* /tests/conftest.py

### CODE REVIEW

**Comment by p0psicles** [comment]
It's not totally working right now. It's hanging on saving the poster/banner, as it's not getting a full url from the tvdb api. But it is adding to db, so it's somewhat testable.

File: medusa/indexers/tvdbv2/tvdbv2_api.py [hunk 1]
```diff
@@ -0,0 +1,877 @@
...
+
+import cgi
+import getpass
+import logging
+import os
+import re
+import tempfile
+import time
+import warnings
+from six import next, iteritems
```
Comment (p0psicles) [inline medusa/indexers/tvdbv2/tvdbv2_api.py]: hmm okay, how should I resolve this? import it like: from six import next as next_six? As i'd like to use a python2/3 compatible method.

File: medusa/indexers/tvmaze/tvmaze_api.py [hunk 2]
```diff
@@ -0,0 +1,816 @@
...
+            if key not in self.shows:
+                self._getShowData(key, self.config['language'], True)
+            return self.shows[key]
+
+        key = str(key).lower()
+        self.config['searchterm'] = key
+        selected_series = self._getSeries(key)
+        if isinstance(selected_series, dict):
+            selected_series = [selected_series]
+        [[self._setShowData(show['id'], k, v) for k, v in show.items()] for show in selected_series]
```
Comment (p0psicles) [inline medusa/indexers/tvmaze/tvmaze_api.py]: I think I can resolv this with _ = [[self._setShowData..., but would that be a proper solution?

File: medusa/indexers/tvdbv2/tvdbv2_api.py [hunk 3]
```diff
@@ -0,0 +1,782 @@
...
+        auth_api = AuthenticationApi(unauthenticated_client)
+        access_token = auth_api.login_post(authentication_string)
+        auth_client = ApiClient(api_base_url, 'Authorization', 'Bearer ' + access_token.token)
+        self.search_api = SearchApi(auth_client)
+        self.series_api = SeriesApi(auth_client)
+
+    def _parse_show_data(self, indexer_data, parsing_into_key='series'):  # pylint: disable=no-self-use
+        """ These are the fields we'd like to see for a show search"""
+
+        name_map = {
```
Comment (p0psicles) [inline medusa/indexers/tvdbv2/tvdbv2_api.py]: This method has been removed/replaced.

**Comment by OmgImAlexis** [comment]
@p0psicles does the new API also allow searching for things like `mr` as well as `mr.`?

**Comment by p0psicles** [comment]
@OmgImAlexis if your referring to mr. robot. No, you still need to search for mr. robot. mr robot, won't work.

File: medusa/indexers/indexer_ui.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
+# coding=utf-8
+# Author: p0psicles
+#
+# This file is part of SickRage.
```
Comment (OmgImAlexis) [inline medusa/indexers/indexer_ui.py]: The whole header needs to be replaced with the Medusa version.

Comment (p0psicles) [inline medusa/indexers/indexer_ui.py]: Yeah, probably more of them need to be fixed. Tnx!

**Comment by p0psicles** [comment]
@pymedusa/developers pls, start testing this over time. As I need to test the new updating.
I've also made some changes to the unicode stuff. So i'm curious how this will work on different setups.

**Comment by fernandog** [comment]
@p0psicles rebase please so we can merge this first after release

**Comment by p0psicles** [comment]
Rebased.

**Comment by fernandog** [comment]
Can you fix so we can merge?

**Pull request merged**